File: frontend/src/stores/customer.ts
import { defineStore } from "pinia";
import { ref } from "vue";
import api from "@/api/client";

// ── 类型 ──

export interface CustomerListItem {
  id: string;
  name: string;
  industry: string | null;
  country: string | null;
  source: string;
  status: string;
  website: string | null;
  contacts_count: number;
  icp_id: string | null;
  icp_name: string | null;
  created_at: string;
  updated_at: string;
}

export interface ContactItem {
  id: string;
  customer_id: string;
  tenant_id: string;
  name: string;
  title: string | null;
  email: string | null;
  phone: string | null;
  linkedin_url: string | null;
  is_primary: boolean;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface CustomerDetail {
  id: string;
  tenant_id: string;
  name: string;
  industry: string | null;
  website: string | null;
  country: string | null;
  city: string | null;
  company_size: string | null;
  description: string | null;
  source: string;
  source_url: string | null;
  icp_id: string | null;
  status: string;
  source_data: Record<string, unknown> | null;
  ai_summary: string | null;
  notes: string | null;
  contacts_count: number;
  contacts: ContactItem[];
  created_at: string;
  updated_at: string;
}

// ── Store ──

export const useCustomerStore = defineStore("customer", () => {
  const list = ref<CustomerListItem[]>([]);
  const total = ref(0);
  const loading = ref(false);
  const current = ref<CustomerDetail | null>(null);
  const searching = ref(false);

  async function fetchList(params?: Record<string, unknown>) {
    loading.value = true;
    try {
      const { data } = await api.get("/customers", { params });
      list.value = data.items;
      total.value = data.total;
    } finally {
      loading.value = false;
    }
  }

  async function fetchDetail(id: string) {
    const { data } = await api.get(`/customers/${id}`);
    current.value = data;
    return data as CustomerDetail;
  }

  async function create(
    formData: Record<string, unknown>
  ): Promise<CustomerDetail> {
    const { data } = await api.post("/customers", formData);
    return data as CustomerDetail;
  }

  async function update(
    id: string,
    formData: Record<string, unknown>
  ): Promise<void> {
    await api.put(`/customers/${id}`, formData);
  }

  async function remove(id: string): Promise<void> {
    await api.delete(`/customers/${id}`);
  }

  async function batchUpdateStatus(
    ids: string[],
    status: string
  ): Promise<number> {
    const { data } = await api.put("/customers/batch/status", { ids, status });
    return data.updated_count;
  }

  async function addContact(
    customerId: string,
    contactData: Record<string, unknown>
  ) {
    const { data } = await api.post(
      `/customers/${customerId}/contacts`,
      contactData
    );
    return data as ContactItem;
  }

  async function updateContact(
    customerId: string,
    contactId: string,
    contactData: Record<string, unknown>
  ) {
    await api.put(
      `/customers/${customerId}/contacts/${contactId}`,
      contactData
    );
  }

  async function removeContact(
    customerId: string,
    contactId: string
  ): Promise<void> {
    await api.delete(`/customers/${customerId}/contacts/${contactId}`);
  }

  async function enrichCustomer(id: string): Promise<void> {
    const token = localStorage.getItem("access_token");
    const baseUrl = import.meta.env.VITE_API_BASE_URL || "/api/v1";

    const response = await fetch(`${baseUrl}/customers/${id}/enrich`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${token}`,
      },
    });

    if (!response.ok) {
      const err = await response.json().catch(() => ({ detail: "请求失败" }));
      throw new Error(err.detail || "补全请求失败");
    }
  }

  async function importExcel(file: File): Promise<{
    created: number;
    skipped: number;
    total: number;
    errors: { row: number; message: string }[];
  }> {
    const formData = new FormData();
    formData.append("file", file);
    const { data } = await api.post("/customers/import", formData, {
      headers: { "Content-Type": "multipart/form-data" },
    });
    return data;
  }

  async function downloadTemplate(): Promise<void> {
    const response = await api.get("/customers/import-template", {
      responseType: "blob",
    });
    const url = window.URL.createObjectURL(new Blob([response.data]));
    const link = document.createElement("a");
    link.href = url;
    link.setAttribute("download", "customer_import_template.xlsx");
    document.body.appendChild(link);
    link.click();
    link.remove();
    window.URL.revokeObjectURL(url);
  }

  async function exportData(params?: Record<string, unknown>) {
    const response = await api.get("/customers/export", {
      params,
      responseType: "blob",
    });
    const url = window.URL.createObjectURL(new Blob([response.data]));
    const link = document.createElement("a");
    link.href = url;
    link.setAttribute("download", "customers.xlsx");
    document.body.appendChild(link);
    link.click();
    link.remove();
    window.URL.revokeObjectURL(url);
  }

  return {
    list, total, loading, current, searching,
    fetchList, fetchDetail, create, update, remove,
    batchUpdateStatus, enrichCustomer, importExcel, downloadTemplate, exportData,
    addContact, updateContact, removeContact,
  };
});

File: backend/app/schemas/customer.py
"""客户管理 Pydantic Schema — Phase 4"""

import uuid
from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, Field


# ── 联系人 ──

class ContactCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    title: Optional[str] = Field(default=None, max_length=255)
    email: Optional[str] = Field(default=None, max_length=255)
    phone: Optional[str] = Field(default=None, max_length=100)
    linkedin_url: Optional[str] = Field(default=None, max_length=512)
    is_primary: bool = False
    contact_type: Optional[str] = Field(default="scraped", max_length=20)
    confidence: Optional[str] = Field(default=None, max_length=10)
    notes: Optional[str] = Field(default=None, max_length=2000)


class ContactUpdateRequest(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=255)
    title: Optional[str] = Field(default=None, max_length=255)
    email: Optional[str] = Field(default=None, max_length=255)
    phone: Optional[str] = Field(default=None, max_length=100)
    linkedin_url: Optional[str] = Field(default=None, max_length=512)
    is_primary: Optional[bool] = None
    contact_type: Optional[str] = Field(default=None, max_length=20)
    confidence: Optional[str] = Field(default=None, max_length=10)
    notes: Optional[str] = Field(default=None, max_length=2000)


class ContactResponse(BaseModel):
    id: uuid.UUID
    customer_id: uuid.UUID
    tenant_id: uuid.UUID
    name: str
    title: Optional[str] = None
    email: Optional[str] = None
    phone: Optional[str] = None
    linkedin_url: Optional[str] = None
    is_primary: bool
    contact_type: Optional[str] = None
    confidence: Optional[str] = None
    notes: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ── 客户 ──

class CustomerCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    industry: Optional[str] = Field(default=None, max_length=100)
    website: Optional[str] = Field(default=None, max_length=255)
    country: Optional[str] = Field(default=None, max_length=100)
    city: Optional[str] = Field(default=None, max_length=100)
    company_size: Optional[str] = Field(default=None, max_length=50)
    description: Optional[str] = Field(default=None, max_length=2000)
    source: str = Field(default="manual", max_length=50)
    source_url: Optional[str] = Field(default=None, max_length=512)
    icp_id: Optional[uuid.UUID] = None
    status: str = Field(default="new", max_length=20)
    notes: Optional[str] = Field(default=None, max_length=2000)
    contacts: Optional[List[ContactCreateRequest]] = None


class CustomerUpdateRequest(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=255)
    industry: Optional[str] = Field(default=None, max_length=100)
    website: Optional[str] = Field(default=None, max_length=255)
    country: Optional[str] = Field(default=None, max_length=100)
    city: Optional[str] = Field(default=None, max_length=100)
    company_size: Optional[str] = Field(default=None, max_length=50)
    description: Optional[str] = Field(default=None, max_length=2000)
    source: Optional[str] = Field(default=None, max_length=50)
    source_url: Optional[str] = Field(default=None, max_length=512)
    icp_id: Optional[uuid.UUID] = None
    status: Optional[str] = Field(default=None, max_length=20)
    source_data: Optional[dict] = None
    ai_summary: Optional[str] = Field(default=None, max_length=5000)
    notes: Optional[str] = Field(default=None, max_length=2000)


class CustomerListItem(BaseModel):
    """列表项（不含大段 JSON）"""
    id: uuid.UUID
    name: str
    industry: Optional[str] = None
    country: Optional[str] = None
    source: str
    status: str
    website: Optional[str] = None
    contacts_count: int = 0
    contacts_with_email_count: int = 0  # 有邮箱地址的联系人数（用于邮件发送筛选）
    icp_id: Optional[uuid.UUID] = None
    icp_name: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class CustomerResponse(BaseModel):
    id: uuid.UUID
    tenant_id: uuid.UUID
    name: str
    industry: Optional[str] = None
    website: Optional[str] = None
    country: Optional[str] = None
    city: Optional[str] = None
    company_size: Optional[str] = None
    description: Optional[str] = None
    source: str
    source_url: Optional[str] = None
    icp_id: Optional[uuid.UUID] = None
    status: str
    source_data: Optional[dict] = None
    ai_summary: Optional[str] = None
    notes: Optional[str] = None
    contacts_count: int = 0
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class CustomerDetailResponse(CustomerResponse):
    """详情（含联系人列表）"""
    contacts: List[ContactResponse] = []


class CustomerListResponse(BaseModel):
    items: List[CustomerListItem]
    total: int
    page: int
    page_size: int


# ── 搜索 ──

class CustomerSearchRequest(BaseModel):
    icp_id: uuid.UUID
    channels: List[str] = Field(default=["google"])
    region: Optional[str] = Field(default=None, max_length=100)


class BatchStatusRequest(BaseModel):
    ids: List[uuid.UUID]
    status: str = Field(min_length=1, max_length=20)


class CustomerImportRowError(BaseModel):
    """Excel 导入单行错误"""
    row: int
    message: str


class CustomerImportResponse(BaseModel):
    """Excel 导入结果"""
    created: int = 0
    skipped: int = 0
    total: int = 0
    errors: List[CustomerImportRowError] = []
